### PR TITLE
Remove DELETE/PUT convention

### PR DESCRIPTION
**Pull request by sileht**
We should just follow restapi.net guidelines.

In particular https://restfulapi.net/idempotent-rest-apis/ HTTP DELETE

That should return 404 when resource does not exists.

### CODE REVIEW

**Comment by sduthil** [comment]
TLDR: I agree that restfulapi.net is a good resource to be linked here. However I disagree on removing the previously defined conventions.

My main interest here is consistency across Wazo Platform REST API. We have had some discussions on minute details of the Wazo Platform API multiple times, and this document attempts to answer the questions that have already been discussed in order to avoid asking the same questions. This document is subject to change of course, if we think it would improve efficiency, simplicity of use or something else.

About DELETE status code:
* the main reason for returning 204 every time is that is makes the client a bit simpler to implement, having one less case to worry about. It is not a very important reason, so my objection is a small one and I would be OK with a DELETE status code 204/404, as long as the behavior is the same in all Wazo Platform REST API.
* In the [idempotence](https://restfulapi.net/idempotent-rest-apis/) section, they do mention DELETE 204/404 status codes, but it does not make a DELETE 204/204 less idempotent: the state change is the same on the server side. So DELETE 204/204 is also a valid choice in this regard. I did not find any other reference of the DELETE 204/404, so I'd say they don't explicitly recommend this over DELETE 204/204. Another objection I have is that this information is not very easy to find. If they had a thorough list of recommended behaviors, I would be very happy with it. But here it is mentioned on the side in a comment that is focused on a general idempotency concept, not on a precise behavior definition, saying "this is the way to go". So in the end, the information is harder to find.

About the PUT response body, I did not find a reference mentioning a recommended behavior. Since the PUT response body behavior is not clearly defined in restfulapi.net, I would like have it written down somewhere in order for everyone to have a precise idea of the conventions used in Wazo Platform REST API.

**Comment by sileht** [comment]
> TLDR: I agree that restfulapi.net is a good resource to be linked here. However I disagree on removing the previously defined conventions.
> 
> My main interest here is consistency across Wazo Platform REST API. We have had some discussions on minute details of the Wazo Platform API multiple times, and this document attempts to answer the questions that have already been discussed in order to avoid asking the same questions. This document is subject to change of course, if we think it would improve efficiency, simplicity of use or something else.
> 
> About DELETE status code:
> 
>     * the main reason for returning 204 every time is that is makes the client a bit simpler to implement, having one less case to worry about. It is not a very important reason, so my objection is a small one and I would be OK with a DELETE status code 204/404, as long as the behavior is the same in all Wazo Platform REST API.
> 
>     * In the [idempotence](https://restfulapi.net/idempotent-rest-apis/) section, they do mention DELETE 204/404 status codes, but it does not make a DELETE 204/204 less idempotent: the state change is the same on the server side. So DELETE 204/204 is also a valid choice in this regard. I did not find any other reference of the DELETE 204/404, so I'd say they don't explicitly recommend this over DELETE 204/204. Another objection I have is that this information is not very easy to find. If they had a thorough list of recommended behaviors, I would be very happy with it. But here it is mentioned on the side in a comment that is focused on a general idempotency concept, not on a precise behavior definition, saying "this is the way to go". So in the end, the information is harder to find.

Just an example of what will happen for our API users, A developer makes a mistake in it's DELETE API CALL (putting wrong ID), our application will not complain about it if we return 204 and the developer will wonder why the data it delete still exists. And it will contact us.
If we return 404, it can debug the thing himself.

All API have ever see return 404 when resource does not exists whatever the HTTP methods in use. If we don't want to do like other, I'm OK, but since people don't read such detail in documentation they will just assume it behave like any other they already use.

I have another design issue like this one of the current APIs, is the "unknown=marshmallow.EXCLUDE" we have everywhere in API.

It allows API users to push extra json data in API calls, that we will ignore, I known the reason is again "makes the client a bit simpler to implement". But the fact is that if a user make a typo in an API call, we will just accept the payload, partially write what he expects in database. Once again it have good change to contact US, to get why it does not works. Instead of been able to debug himself.

> About the PUT response body, I did not find a reference mentioning a recommended behavior. Since the PUT response body behavior is not clearly defined in restfulapi.net, I would like have it written down somewhere in order for everyone to have a precise idea of the conventions used in Wazo Platform REST API.

The main reason why PUT must contains the body is to avoid API user to have to do two API calls to known what have been changed.

POST-PUT-PATCH and GET are not symmetric. For example, when you link two resources, in POST/PUT you just put ID, while the GET returns the whole resources for this IDs.



All of this is my 2cents, if "consistency" is the master word here, so we can just close this PR. (And  fix swarm-subscription and all C4 codes to match this spec).

Otherwise, I would prefer "consistency" within the same services, new code will use new specs, but old will still continue to follow the old one until we migrate the whole service or have to rewrite a new version of the API.
